source/_posts/2017-08-11-How-we-debug-with-Charles.md

```diff
@@ -0,0 +1,103 @@
+
+
+
+}![rewrite1](https://cdn-laravel.vapor.cloud/image/nstack/translate_values/tool_rewrite1_3CeaIAAvHJ.png)
+
+_First we create a rewrite rule inorder to replace the URL of the iPhone X image_
+
+
+![rewrite2](https://cdn-laravel.vapor.cloud/image/nstack/translate_values/tool_rewrite2_hljikhsImW.jpeg)
+
+_We then reopen the app to force the app to make the request again_
+
+
+![rewrite3](https://cdn-laravel.vapor.cloud/image/nstack/translate_values/tool_rewrite3_UynlzcU5tw.png)
+
+_Checking in our session, we can now see that the rewrite rule has been applied to the request_
+
+
+All request can be applied the rewrite functionality, and multiple rewrites can be configured at the same time. Rewrites are encapsulated into rule sets, which can be saved in your Charles config for future use. Rewrite rules can be disabled and enabled as you please along the testing. As you can see, the options of rewriting are many. There are various reasons for why you might need to add rewrite rules, but these are the most common use cases for us:
+
+* Rewriting status codes
+* Rewriting API url
+* Rewriting authentication headers
+* Rewriting headers to alter user agent
+* Injecting extra headers
+* Injecting response headers
+* Adding a find-and-replace rule on the body of a request or the response
+
+
+### 4. Simulating API responses
+
+With the 'Map local' tool testers have the ability to modify the responses on the fly. The 'Map local' tool changes the response of the request, per the configured mappings, so that the response is transparently served from the new location locally as if that was the original request. In other words the map local will allow testers to simulate responses from each request. We use this tool to quickly modify the response body locally in order to test how app reacts to it. The simplest example could be modifying a response from an API call to see if the app UI can handle longer text values. Another use case, is when testers or developers are not getting the expected response from the API either due to new features not being deployed or bugs in the API. In that case we can modify the API response locally to the expected in order to be able to continue testing or developing. We often see that logic, which depends on certain properties/values in the response from the API, is difficult to test for. Sometimes testers will need to go through complicated flows in the app in order to have the response properties updated accordingly. Here the map local comes in handy, as we can quickly modify the response and change the required parameter to what we need, in order to be able to test.
+
```

Comment:
> With the 'Map local' tool testers

, between tool and testers so "With the 'Map local' tool, testers"

Reply:
> We use this tool to quickly modify

New paragraph before this line

Reply:
>  In that case we can modify the API response locally to the expected in order to be able to continue testing or developing.

I would change it to " In that case we can modify the API response locally **to what we would expect it to be so we can continue testing or developing**"

Reply:
> Sometimes testers

"Sometimes, testers"

and maybe a new paragraph before as well

Reply:
> Here the map local 

"Here, the map local"